### ISSUE TITLE
Deal with options_.nobs and options_.first_obs in master

### ISSUE THREAD

**Issue by JohannesPfeifer** [DynareTeam/dynare#780]
With the new data interface, these fields are not always set, leading to crashes in different places. For example, recursive forecasting still relies on those fields. See also afb5be206741457a9b6383fa004bb3f208ed5d94 
We need to decide on how to treat them. Are these field properties of the dataset to be moved into `dataset_`? Or are they to be used as options in other commands?

**Comment by JohannesPfeifer** [comment]
@stepan-a I understand that you have a lot to deal with, but my tickets rely on this bug being fixed. If you find 5 minutes, could you please just fix this along the lines outlined above.

**Comment by stepan-a** [comment]
@JohannesPfeifer, I still have a problem with your proposed change. I tried to run a recursive estimation with fs2000.mod with the old data interface and it ran perfectly fine... So I do not understand the problem you are trying to fix. Can you point me to a mod file where the issue you are referring pops up?

**Comment by JohannesPfeifer** [comment]
@stepan-a The issue in recursive estimation is gone by now. What remains is the problem in e.g. ```dyn_forecast``` and ```filt_mc_```. Just take the fs2000.mod and add the ```forecast``` option: 
```
/*
 * This file replicates the estimation of the cash in advance model described
 * Frank Schorfheide (2000): "Loss function-based evaluation of DSGE models",
 * Journal of Applied Econometrics, 15(6), 645-670.
 *
 * The data are in file "fsdat_simul.m", and have been artificially generated.
 * They are therefore different from the original dataset used by Schorfheide.
 *
 * The equations are taken from J. Nason and T. Cogley (1994): "Testing the
 * implications of long-run neutrality for monetary business cycle models",
 * Journal of Applied Econometrics, 9, S37-S70.
 * Note that there is an initial minus sign missing in equation (A1), p. S63.
 *
 * This implementation was written by Michel Juillard. Please note that the
 * following copyright notice only applies to this Dynare implementation of the
 * model.
 */

/*
 * Copyright (C) 2004-2010 Dynare Team
 *
 * This file is part of Dynare.
 *
 * Dynare is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Dynare is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Dynare.  If not, see <http://www.gnu.org/licenses/>.
 */

var m P c e W R k d n l gy_obs gp_obs y dA;
varexo e_a e_m;

parameters alp bet gam mst rho psi del;

alp = 0.33;
bet = 0.99;
gam = 0.003;
mst = 1.011;
rho = 0.7;
psi = 0.787;
del = 0.02;

model;
dA = exp(gam+e_a);
log(m) = (1-rho)*log(mst) + rho*log(m(-1))+e_m;
-P/(c(+1)*P(+1)*m)+bet*P(+1)*(alp*exp(-alp*(gam+log(e(+1))))*k^(alp-1)*n(+1)^(1-alp)+(1-del)*exp(-(gam+log(e(+1)))))/(c(+2)*P(+2)*m(+1))=0;
W = l/n;
-(psi/(1-psi))*(c*P/(1-n))+l/n = 0;
R = P*(1-alp)*exp(-alp*(gam+e_a))*k(-1)^alp*n^(-alp)/W;
1/(c*P)-bet*P*(1-alp)*exp(-alp*(gam+e_a))*k(-1)^alp*n^(1-alp)/(m*l*c(+1)*P(+1)) = 0;
c+k = exp(-alp*(gam+e_a))*k(-1)^alp*n^(1-alp)+(1-del)*exp(-(gam+e_a))*k(-1);
P*c = m;
m-1+d = l;
e = exp(e_a);
y = k(-1)^alp*n^(1-alp)*exp(-alp*(gam+e_a));
gy_obs = dA*y/y(-1);
gp_obs = (P/P(-1))*m(-1)/dA;
end;

initval;
k = 6;
m = mst;
P = 2.25;
c = 0.45;
e = 1;
W = 4;
R = 1.02;
d = 0.85;
n = 0.19;
l = 0.86;
y = 0.6;
gy_obs = exp(gam);
gp_obs = exp(-gam);
dA = exp(gam);
end;

shocks;
var e_a; stderr 0.014;
var e_m; stderr 0.005;
end;

steady_state_model;
  dA = exp(gam);
  gst = 1/dA;
  m = mst;
  khst = ( (1-gst*bet*(1-del)) / (alp*gst^alp*bet) )^(1/(alp-1));
  xist = ( ((khst*gst)^alp - (1-gst*(1-del))*khst)/mst )^(-1);
  nust = psi*mst^2/( (1-alp)*(1-psi)*bet*gst^alp*khst^alp );
  n  = xist/(nust+xist);
  P  = xist + nust;
  k  = khst*n;

  l  = psi*mst*n/( (1-psi)*(1-n) );
  c  = mst/P;
  d  = l - mst + 1;
  y  = k^alp*n^(1-alp)*gst^alp;
  R  = mst/bet;
  W  = l/n;
  ist  = y-c;
  q  = 1 - d;

  e = 1;
  
  gp_obs = m/dA;
  gy_obs = dA;
end;

steady;

check;

estimated_params;
alp, beta_pdf, 0.356, 0.02;
bet, beta_pdf, 0.993, 0.002;
gam, normal_pdf, 0.0085, 0.003;
mst, normal_pdf, 1.0002, 0.007;
rho, beta_pdf, 0.129, 0.223;
psi, beta_pdf, 0.65, 0.05;
del, beta_pdf, 0.01, 0.005;
stderr e_a, inv_gamma_pdf, 0.035449, inf;
stderr e_m, inv_gamma_pdf, 0.008862, inf;
end;

varobs gp_obs gy_obs;

estimation(order=1, datafile=fsdat_simul, nobs=192, loglinear, mh_replic=2000, mh_nblocks=1, mh_jscale=0.8,forecast=8);

```
 You will see that the forecast graph does not show any lines. This is due to ```nobs``` being NaN.

**Comment by JohannesPfeifer** [comment]
@stepan-a This also breaks the ```ppost``` option of ```dynare_sensitivity``` which relies on ```options_.nobs``` being always set.

**Comment by JohannesPfeifer** [comment]
@MichelJuillard https://github.com/DynareTeam/dynare/commit/eb3d0cd20e0d061a05af0ed0f0453003eacc5064 maybe solves this issue, but @stepan-a might disagree with the commit.

**Comment by MichelJuillard** [comment]
We need to proceed by steps. options_.nobs is still needed by some part of the codes. Commit
https://github.com/DynareTeam/dynare/commit/eb3d0cd20e0d061a05af0ed0f0453003eacc5064 only makes sure that it is correctly set and corrects a bug in  dynare_estimation_init.m
If it turns out that options_.nobs is not desirable anymore, the change should be made systematically in all functions using it.

**Comment by JohannesPfeifer** [comment]
Seems to be solved with #1412
